File: Top_100_Liked_Questions/19._Remove_Nth_Node_From_End_of_List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        # two pointer with constant gap method
        dummyHead = ListNode(0, head)
        fast = slow = dummyHead
        for _ in range(n):
            fast = fast.next
        
        while fast.next:
            fast = fast.next
            slow = slow.next
      
        delNode = slow.next
        slow.next = delNode.next
        delNode.next = None
        del delNode

        head = dummyHead.next
        dummyHead.next = None
        del dummyHead
        
        return head

File: Top_100_Liked_Questions/19._Remove_Nth_Node_From_End_of_List/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let dummyHead = new ListNode(0, head);
    let slow = dummyHead;
    let fast = dummyHead;
    
    for (let i=0; i<n; i++) {
        fast = fast.next;
    }
    
    while (fast.next) {
        fast = fast.next;
        slow = slow.next;
    }
    
    let delNode = slow.next;
    slow.next = delNode.next;
    delNode.next = null;
    delete delNode;
    
    head = dummyHead.next;
    dummyHead.next = null;
    delete dummyHead;
    
    return head;
};